### PR TITLE
refactor gateway-service, add player smooth movement, refactor input manager

### PR DESCRIPTION
**Pull request by ktpss95112**
Still working in progress
9/24 updated: done

### CODE REVIEW

File: common/gamelib/move-check.mjs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+const PLAYER_MOVE_DISTANCE_LIMIT = 1; // cell
+
+/**
+ * TODO
+ * @param {Player} oldPlayerData - TODO
+ * @param {PlayerSyncMessage} newPlayerData - TODO
+ * @param {GameMap} gameMap - TODO
+ * @return {Boolean}
+ */
+function checkPlayerMove(oldPlayerData, newPlayerData, gameMap) {
```
Comment (john0312) [inline common/gamelib/move-check.mjs:14]: It seems that `newPlayerData` is `PlayerSyncMessage` type while `oldPlayerData` is `Player` type. Maybe we can rename the variable for better readability?

File: common/gamelib/player.mjs [hunk 2]
```diff
@@ -57,11 +85,15 @@ class Player {
...
-    if (msg.mapCoord !== undefined) this.mapCoord = msg.mapCoord;
+    if (msg.mapCoord !== undefined) {
+      this._previousMapCoord = (this.mapCoord ?? msg.mapCoord); // for smoothly move the player
+      this.mapCoord = msg.mapCoord;
+    }
     if (msg.facing !== undefined) this.facing = msg.facing;
     if (msg.displayName !== undefined) this.displayName = msg.displayName;
     if (msg.displayChar !== undefined) this.displayChar = msg.displayChar;
     if (msg.removed !== undefined) this.removed = msg.removed;
+    this.lastMovingTime = Date.now();
```
Comment (john0312) [inline common/gamelib/player.mjs:96]: Maybe set `this.lastMovingTime` only if `this.mapCoord` changed? Considering `updateFromMessage()` is not limited to only moving the user.

File: services/gateway/gateway-service.mjs [hunk 3]
```diff
@@ -235,57 +236,28 @@ class GatewayService {
...
-    if (socket.playerID !== origMsg.playerID) {
-      console.error(`Player '${origMsg.playerID}' tries to update player '${socket.playerID}'s data, which is invalid.`);
+  async onPlayerUpdate(socket, updateMsg) {
+    if (socket.playerID !== updateMsg.playerID) {
+      console.error(`Player '${updateMsg.playerID}' tries to update player '${socket.playerID}'s data, which is invalid.`);
       return;
     }
 
-    const msg = PlayerSyncMessage.fromObject(socket.playerData);
-    msg.facing = origMsg.facing;
```
Comment (john0312) [inline services/gateway/gateway-service.mjs:248]: I think it makes sense to explicitly copy over the fields that we want to allow the users to update (the way it's done in the deleted code). This way, whenever a new field is added, there's lesser chance for us to introduce logical bug that allows user to manipulate the field without check.

**Review by john0312 — APPROVED** [review]

File: services/gateway/gateway-service.mjs (hunk 3 above)
Comment (ktpss95112) [reply services/gateway/gateway-service.mjs:248]: One of the purposes to use `PlayerSyncMessage` class is to restrict the fields that can be updated. I think it is unnecessary for the server to construct the update message again.

File: common/gamelib/player.mjs (hunk 2 above)
Comment (ktpss95112) [reply common/gamelib/player.mjs]: fixed

File: common/gamelib/move-check.mjs (hunk 1 above)
Comment (ktpss95112) [reply common/gamelib/move-check.mjs]: fixed

**Pull request merged**